src/backend/distributed/master/master_truncate.c

```diff
@@ -73,7 +73,15 @@ citus_truncate_trigger(PG_FUNCTION_ARGS)
 	{
 		List *taskList = TruncateTaskList(relationId);
 
-		ExecuteUtilityTaskListWithoutResults(taskList);
+		/*
+		 * If it is a local placement of a distributed table or a reference table,
+		 * then execute TRUNCATE command locally.
+		 */
+		bool localExecutionSupported = true;
+
+		UseCoordinatedTransaction();
```

Comment:
> Overall, calling ShouldExecuteTasksLocally multiple times seems the root cause.

We shouldn't avoid this, we should aim to make  `ShouldExecuteTasksLocally` idempotent

Reply:
Makes sense to me. But I wonder if we really should not use coordinated transaction for truncate commands as well ?

Reply:
I wonder if we really shouldn't execute TRUNCATE commands in coordinated transactions as we do for the other type of commands

Reply:
For local only execution, we don't need to be in `UseCoordinatedTransaction`. Coordinated transaction is only useful when the coordinator starts remote transactions on the workers. See `StartDistributedExecution` for some details.

Reply:
Other than this, I purpose to remove that assertion since as you also mentioned,  `ShouldExecuteTasksLocally` can also be called via triggers (starting with my patch) and those kind of assertion exceptipns could be inevitable in those cases

Reply:
As we discussed in person, let's try to get rid of this `ExecuteLocalUtilityTaskList ` and integrate the logic in `ExecuteTaskList `. That'd solve this problem and make the code simpler.

Reply:
The following is a patch, that could allow us to remove `UseCoordinatedTransaction()` here and make the `Assert` explicitly consider this case. This is a new case, and we could consider making it explicit here, instead  of removing all the Asserts.

What do you think @onurctirtir: 
```C
diff --git a/src/backend/distributed/executor/local_executor.c b/src/backend/distributed/executor/local_executor.c
index 2a418927e..4f02466e9 100644
--- a/src/backend/distributed/executor/local_executor.c
+++ b/src/backend/distributed/executor/local_executor.c
@@ -532,16 +532,32 @@ ShouldExecuteTasksLocally(List *taskList)
 
 	if (TransactionAccessedLocalPlacement)
 	{
+		bool localExecutionExpected PG_USED_FOR_ASSERTS_ONLY = false;
+
 		/*
 		 * For various reasons, including the transaction visibility
 		 * rules (e.g., read-your-own-writes), we have to use local
 		 * execution again if it has already happened within this
 		 * transaction block.
-		 *
+		 */
+		localExecutionExpected =
+			IsMultiStatementTransaction() || InCoordinatedTransaction();
+
+		/*
+		 * In some cases, such as when a single command leads to a
+		 * local command execution followed by remote task (list)
+		 * execution, we still expect the remote execution to first try
+		 * local execution as TransactionAccessedLocalPlacement is set.
+		 * In the end, the remote execution wouldn't create any local
+		 * tasks, and we're ensuring it here.
+		 */
+		localExecutionExpected |= !AnyTaskAccessesLocalNode(taskList);
+
+		/*
 		 * We might error out later in the execution if it is not suitable
 		 * to execute the tasks locally.
 		 */
-		Assert(IsMultiStatementTransaction() || InCoordinatedTransaction());
+		Assert(localExecutionExpected);
 
 		/*
 		 * TODO: A future improvement could be to keep track of which placements
```

Reply:
seems good to me, I should try it on enabling assertions to see if it works